state/watcher.go

```diff
@@ -496,6 +497,154 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachineStartTimes watches the non-container machines in the model
+// for changes to the Life or AgentStartTime fields and reports them as a batch
+// after the specified quiesceInterval time has passed without seeing any new
+// change events.
+func (st *State) WatchModelMachineStartTimes(quiesceInterval time.Duration) StringsWatcher {
+	return newModelMachineStartTimeWatcher(st, st.clock(), quiesceInterval)
+}
+
+type modelMachineStartTimeWatcher struct {
+	commonWatcher
+	outCh chan []string
+
+	clk             clock.Clock
+	quiesceInterval time.Duration
+	docHash         map[string]string
+}
+
+func newModelMachineStartTimeWatcher(backend modelBackend, clk clock.Clock, quiesceInterval time.Duration) StringsWatcher {
+	w := &modelMachineStartTimeWatcher{
+		commonWatcher:   newCommonWatcher(backend),
+		outCh:           make(chan []string),
+		clk:             clk,
+		quiesceInterval: quiesceInterval,
+		docHash:         make(map[string]string),
+	}
+
+	w.tomb.Go(func() error {
+		defer close(w.outCh)
+		return w.loop()
+	})
+	return w
+}
+
+// SetClock replaces the clock implementation used by the watcher. This method
+// is meant to be used by tests.
+func (w *modelMachineStartTimeWatcher) SetClock(clk clock.Clock) {
```

Comment:
I think it doesn't because the tests do not share the watcher. I used this approach because I didn't want to add a new c-tor (nothing else uses this pattern in that file) or do the add-specialized-ctor-in-exports hack.

However, it turns out that with a careful combination of WaitForModelWatchersIdle in the test I can simply advance the testclock used by the suite without the need to use WaitAndAdvance! I have refactored my test and removed this method (verified it passes -race -count 30 with no issues.)

Reply:
It would still be a race even without the watcher being shared between tests, right? The main goroutine is writing to the attribute while the loop reads from it.

It wouldn't have to use the exports hack necessarily - you've got a public method to set the clock from tests, I'd be fine with a public alternative constructor that only tests used. But if you've worked out a way to test without the clock, nice.